### PR TITLE
Validation/trainsplit

### PR DESCRIPTION
**Pull request by HMurawski**
Added: 
train.py - splitting the set to train and validation (45000/5000), storing the best outcomes after training - monitoring val_loss and val_accuracy after each epoch to detect potential overfitting instead of immediately using the test set
test.py - deleted the range(5) loop from closing the visualization, making the testing on the best model that was being saved during training to get a proper final acc score.

### CODE REVIEW

**Comment by jrojewski** [comment]
I checked it and It works fine for me 🎈

File: test.py [hunk 1]
```diff
@@ -1,11 +1,15 @@
 import torch
 import torchvision.transforms as transforms
 from torchvision.datasets import CIFAR10
+from torch.utils.data import DataLoader
 from models.simple_cnn import SimpleCNN
 from utils import imshow
 import torchvision
 
 def main():
+    # setting up GPU as default device
```
Comment (lzdunowski) [inline test.py:10]: 'setting up CUDA as default device'*
as other GPUs not included

File: test.py [hunk 2]
```diff
@@ -14,37 +18,46 @@ def main():
...
+    model = SimpleCNN().to(device)
+    model.load_state_dict(torch.load("best_model.pth", map_location=device))
     model.eval()
 
     # Displaying sample images with predictions
     dataiter = iter(testloader)
     images, labels = next(dataiter)
+    
+    images = images.to(device)
+    labels = labels.to(device)
```
Comment (lzdunowski) [inline test.py:33]: Wouldn't extending line 30 be sufficient?
"images, labels = next(dataiter).to(device)"

File: test.py [hunk 3]
```diff
@@ -14,37 +18,46 @@ def main():
...
     outputs = model(images)
     _, predicted = torch.max(outputs, 1)
-
-    for i in range(5):
-        imshow(torchvision.utils.make_grid(images))
-
-    print("Rzeczywiste etykiety(test):", labels.numpy())
-    print("Przewidywane etykiety(test):", predicted.numpy())
+    
+    #transfering back to CPU for visualization purposes
```
Comment (lzdunowski) [inline test.py:38]: Why back to CPU instead of keeping chosen device?

File: test.py [hunk 4]
```diff
@@ -14,37 +18,46 @@ def main():
...
     with torch.no_grad():
         for inputs, labels in testloader:
+            inputs = inputs.to(device)
+            labels = labels.to(device)
             outputs = model(inputs)
             _, predicted = torch.max(outputs, 1)
             total += labels.size(0)
             correct += (predicted == labels).sum().item()
+            
+    accuracy = 100.0 * correct / total
```
Comment (lzdunowski) [inline test.py:57]: Do we need additional variable here?

File: test.py [hunk 5]
```diff
@@ -14,37 +18,46 @@ def main():
...
     # load testing data
     testset = CIFAR10(root='./data', train=False, download=True, transform=transform)
-    testloader = torch.utils.data.DataLoader(testset, batch_size=64, shuffle=False)
+    testloader = DataLoader(testset, batch_size=64, shuffle=False)
 
     # Read model
-    model = SimpleCNN()
-    model.load_state_dict(torch.load("model.pth"))
+    model = SimpleCNN().to(device)
+    model.load_state_dict(torch.load("best_model.pth", map_location=device))
```
Comment (lzdunowski) [inline test.py:25]: Just wondering if 'model' object (line 24) is already created with '.to(device)',
it is needed to use map_location=model?

File: train.py [hunk 6]
```diff
@@ -2,10 +2,16 @@
 import torchvision
 import torchvision.transforms as transforms
 from torch import optim, nn
+from torch.utils.data import random_split, DataLoader
 from models.simple_cnn import SimpleCNN
 from utils import imshow
+import matplotlib.pyplot as plt
```
Comment (lzdunowski) [inline train.py:8]: Do we need it?

File: train.py [hunk 7]
```diff
@@ -2,10 +2,16 @@
 import torchvision
 import torchvision.transforms as transforms
 from torch import optim, nn
+from torch.utils.data import random_split, DataLoader
 from models.simple_cnn import SimpleCNN
 from utils import imshow
+import matplotlib.pyplot as plt
 
 def main():
+    # setting up GPU as default device
```
Comment (lzdunowski) [inline train.py:11]: Same as in 'test.py'.

File: train.py [hunk 8]
```diff
@@ -26,26 +40,71 @@ def main():
 
     # model init
     print("------init model  ------")
-    model = SimpleCNN()
+    model = SimpleCNN().to(device) #transfering to GPU
```
Comment (lzdunowski) [inline train.py:43]: Comment not needed as it might be transfered to GPU or not.

File: train.py [hunk 9]
```diff
@@ -16,7 +22,15 @@ def main():
 
     # Loading data
     trainset = torchvision.datasets.CIFAR10(root='./data', train=True, download=True, transform=transform)
-    trainloader = torch.utils.data.DataLoader(trainset, batch_size=64, shuffle=True)
+    
+    # 4. splitting the set to get train+validation
+    train_size = 45000
+    val_size = 5000
```
Comment (lzdunowski) [inline train.py:28]: 45000 and 5000 are just random numbers or is there any reason of picking it?

File: train.py [hunk 10]
```diff
@@ -26,26 +40,71 @@ def main():
...
+        #VALIDATION
+        model.eval()
+        val_loss = 0.0
+        correct = 0
+        total = 0
+        
+        with torch.no_grad():
+            for inputs, labels in valloader:
+                inputs = inputs.to(device)
+                labels = labels.to(device)
```
Comment (lzdunowski) [inline train.py:78]: just a little thought - I would consider if transfer .to(device) (gpu in that case) needs to be perform here every loop run. Maybe some other option/place to improve performance.

File: train.py [hunk 11]
```diff
@@ -26,26 +40,71 @@ def main():
...
+        if val_acc > best_val_acc:
+            best_val_acc = val_acc
+            torch.save(model.state_dict(), "best_model.pth")
+            print("Model improved - saved to best_model.pth")
+        
+    #Saving the last state
+    torch.save(model.state_dict(), "model_last.pth")
+    print("Training finished! ")
+    print(f"Best Val Accuracy: {best_val_acc:.2f}%")
+
```
Comment (lzdunowski) [inline train.py:105]: Overall looks good :)

**Review by lzdunowski — CHANGES_REQUESTED** [review]

File: test.py (hunk 3 above)
Comment (HMurawski) [reply test.py:38]: matplotlib doesnt support cuda for visualization

File: train.py (hunk 8 above)
Comment (HMurawski) [reply train.py:43]: agreed

File: train.py (hunk 9 above)
Comment (HMurawski) [reply train.py:28]: just a first try to see the outcome, numbers may and should be fixed to resolve the best pairing

File: test.py (hunk 2 above)
Comment (HMurawski) [reply test.py:33]: wont work, since .to(device) is for torch.Tensor objects and next(dataiter) is returning a tuple. Shorter, possible solution would be: 
images, labels = next(dataiter)
images, labels = images.to(device), labels.to(device)
(doesnt matter imo, but suggested solution wont work sadly)

File: test.py (hunk 4 above)
Comment (HMurawski) [reply test.py:57]: wdym? i think is the only variable to see the acc in test.py, unless im wrong?

File: test.py (hunk 5 above)
Comment (HMurawski) [reply test.py:25]: probably not necessary, but using it is preffered according to best practices and to avoid using wrong device by pytorch(or at least I think so  👯 )

File: train.py (hunk 6 above)
Comment (HMurawski) [reply train.py:8]: nope, free to being deleted

File: train.py (hunk 10 above)
Comment (HMurawski) [reply train.py:78]: yup, due to transfering model to gpu, we must transfer it to gpu in every iteration, since DataLoader is returning new batch(inputs and labels) every time at the cpu as default. Your solution could work if we would be able to put all the dataset into GPU, but in our case we are using  batch_size=64
